lib/migration/manager/manager.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/version"
+	"github.com/vmware/vic/pkg/vsphere/session"
+)
+
+const (
+	ApplianceConfigure = "ApplianceConfigure"
+	ContainerConfigure = "ContainerConfigure"
+
+	ApplianceVersionKey = "guestinfo.vice./init/version/PluginVersion"
+	ContainerVersionKey = "guestinfo.vice./version/PluginVersion"
+)
+
+var (
+	Migrator = NewDataMigrator()
+)
+
+type Plugin interface {
+	Migrate(ctx context.Context, s *session.Session, data interface{}) error
+}
+
+type DataMigration interface {
+	// Register plugin to data migration system
+	Register(version int, target string, plugin Plugin) error
+	// Migrate data with current version ID, return true if has any plugin executed
+	Migrate(ctx context.Context, s *session.Session, target string, currentVersion int, data interface{}) (int, error)
+	// GetLatestVersion return the latest plugin id for specified target
+	GetLatestVersion(target string) int
+}
+
+type DataMigrator struct {
+	targetVers map[string][]int
+	verPlugins map[int]Plugin
+}
+
+func NewDataMigrator() DataMigration {
+	return &DataMigrator{
+		targetVers: make(map[string][]int),
+		verPlugins: make(map[int]Plugin),
+	}
+}
+
+func (m *DataMigrator) Register(ver int, target string, plugin Plugin) error {
+	defer trace.End(trace.Begin(fmt.Sprintf("plugin %s:%d", target, ver)))
+	// assert if plugin version less than max plugin version, which is forcing deveoper to change MaxPluginVersion variable everytime new plugin is added
+	if plugin == nil {
+		return &errors.InternalError{
+			Message: "Empty Plugin object is not allowed",
+		}
+	}
+	if ver > version.MaxPluginVersion {
+		return &errors.InternalError{
+			Message: fmt.Sprintf("Plugin %d is bigger than Max Plugin Version %d", ver, version.MaxPluginVersion),
+		}
+	}
+
+	if m.verPlugins[ver] != nil {
+		return &errors.InternalError{
+			Message: fmt.Sprintf("Plugin %d is conflict with another plugin, please make sure the plugin Version is unique and ascending", ver),
+		}
+	}
+
+	m.insertVersion(ver, target)
+	m.verPlugins[ver] = plugin
+	return nil
+}
+
+func (m *DataMigrator) insertVersion(version int, target string) {
+	defer trace.End(trace.Begin(fmt.Sprintf("insert %s:%d", target, version)))
+
+	s := m.targetVers[target]
+	if len(s) == 0 {
+		m.targetVers[target] = append(s, version)
+		return
+	}
+	i := sort.SearchInts(s, version)
+	m.targetVers[target] = append(s[:i], append([]int{version}, s[i:]...)...)
+	log.Debugf("version array: %d", m.targetVers[target])
+}
+
+func (m *DataMigrator) Migrate(ctx context.Context, s *session.Session, target string, currentVersion int, data interface{}) (int, error) {
+	defer trace.End(trace.Begin(fmt.Sprintf("migrate %s from %d", target, currentVersion)))
+
+	pluginVers := m.targetVers[target]
+	if len(pluginVers) == 0 {
+		log.Debugf("No plugin registered for %s", target)
+		return currentVersion, nil
+	}
+
+	i := sort.SearchInts(pluginVers, currentVersion)
```

Comment:
This is a little spooky part here. PluginVers supposed to be in ascending sorted order. I see no reference to that anywhere in docs around here.

I think it would be clear if you just built and a new array of ints which satisfy expected criteria. So you would even eliminate bunch of logic lower.